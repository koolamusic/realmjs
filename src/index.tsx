import React from 'react'


const Realm: React.FC = () => {
    return (
        <div>I am a realm</div>
    )
}

export default Realm