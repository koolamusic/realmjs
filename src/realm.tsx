import { useState, useMemo } from 'react';
import { useEffect } from 'react';


type TValue = string | string[] | number[] | undefined | number | boolean


export interface RealmStoreProps<T extends TValue> {
    get(): T;
    set: (newValue: T) => void;
    subscribe: (listenerFunc: Function) => void;
}

/* -------------------------------------------------------------------------- */
/*                             useMount                                       */
/* -------------------------------------------------------------------------- */

type TMountCallback = { <T>(): T, (): void }

const useMount = (cb: TMountCallback) =>
    useEffect(() => {
        cb();
    }, []);



/* -------------------------------------------------------------------------- */
/*                             useRealm                                       */
/* -------------------------------------------------------------------------- */
export const useRealm = <T extends TValue>(store: RealmStoreProps<T>): [T, Function] => {
    const [state, setGlobal] = useState<T>(store.get());

    useMount(() => {
        return store.subscribe(setGlobal);
    });

    const setState = useMemo(() => (state: T) => store.set(state), [store]);
    return [state, setState];
};





/* -------------------------------------------------------------------------- */
/*                             createRealm                                    */
/* -------------------------------------------------------------------------- */

export const createRealm = <T extends TValue>(initialData: T): RealmStoreProps<T> => {
    let listeners: Function[] = [];
    let value = initialData;

    /* Get Realm Value */
    const get = () => {
        return value;
    };

    /* Set Realm Value */
    const set = (newValue: T) => {
        if (value === newValue) return;
        value = newValue;
        listeners.forEach(listener => listener(value));
    };

    /* Unsubscribe from global listener */
    const unsubscribe = (listenerFunc: Function) => {
        listeners = listeners.filter(listener => listener !== listenerFunc);
    };

    /* Subscribe to global listener */
    const subscribe = (listenerFunc: Function) => {
        listeners.push(listenerFunc);
        return () => unsubscribe(listenerFunc);
    };

    return {
        get,
        set,
        subscribe,
    }
};
