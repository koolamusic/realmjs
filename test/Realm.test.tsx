import React from 'react';
import { useRealm, createRealm } from '../src/index';
import renderer from 'react-test-renderer';


export const COUNTER = createRealm<number>(0);
const AppWithRealm = () => {
    const [state, setState] = useRealm<number>(COUNTER);
    return (
        <React.Fragment>
            <h1>Use Realm Example</h1>
            <h3>{state}</h3>
            <section>
                <button onClick={() => setState(state + 1)}>+</button>
                <button onClick={() => setState(state - 1)} disabled={state === 0}>-</button>
            </section>
        </React.Fragment>
    );
}
describe('realm ', () => {
    test('it renders', () => {

        const tree = renderer.create(<AppWithRealm />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
