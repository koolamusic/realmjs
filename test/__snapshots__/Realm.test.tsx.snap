// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`realm  it renders 1`] = `
Array [
  <h1>
    Use Realm Example
  </h1>,
  <h3>
    0
  </h3>,
  <section>
    <button
      onClick={[Function]}
    >
      +
    </button>
    <button
      disabled={true}
      onClick={[Function]}
    >
      -
    </button>
  </section>,
]
`;
